### PR TITLE
Stream logging to kibana

### PR DESCRIPTION
**Pull request by jam-jam**
1 logFile per job.
If debug, refused files logged as well. 
All logs also sent to logstash if elk.destination defined.

### CODE REVIEW

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java [hunk 1]
```diff
@@ -97,19 +102,33 @@
...
+		encounteredExcludedLinks = new HashSet<String>();
 		JobDataMap dataMap = context.getJobDetail().getJobDataMap();
-		Consumer<Map<String, Object>> consumer = (Consumer<Map<String, Object>>) dataMap.get(WebFetcher.PROPERTY_CONSUMER);
+		Consumer<Map<String, Object>> consumer = (Consumer<Map<String, Object>>) dataMap
+				.get(WebFetcher.PROPERTY_CONSUMER);
 		String url = dataMap.getString(WebFetcher.PROPERTY_URL);
+		String logFileName = setLogFileName(url);
+		jobName = logFileName;
+		// MDC.put("logFileNameLinks",logFileName+"_excludedLinks");
+		// MDC.put("logFileNameData",logFileName+"_excludedData");
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java]: dont just comment out code

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java [hunk 2]
```diff
@@ -97,19 +102,33 @@
...
+	private String jobName = "";
 
 	private List<String> tmpList = new ArrayList<>();
 
 	@Override
 	public void execute(JobExecutionContext context) throws JobExecutionException {
 
+		encounteredExcludedLinks = new HashSet<String>();
 		JobDataMap dataMap = context.getJobDetail().getJobDataMap();
-		Consumer<Map<String, Object>> consumer = (Consumer<Map<String, Object>>) dataMap.get(WebFetcher.PROPERTY_CONSUMER);
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java]: no need to change the formatting

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java [hunk 3]
```diff
@@ -164,10 +187,26 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
...
 		}
 
-		LOGGER.info("Finished Thread {}", threadId);
+		logJob(url, maxPagesCount, encounteredExcludedLinks, excludedDataPagesCount, urlsForDeletionCount);
 
 	}
 
+
+	private String setLogFileName(String url) {
+		// TODO Auto-generated method stub
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java]: remove this

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java [hunk 4]
```diff
@@ -164,10 +187,26 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
...
+		String logFileName = url.replace("http://europa.eu/", "");
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java]: we cant have europa specific login in a public project

**Comment by jam-jam** [comment]
Added _logFileName_ as a param in the webfetcher config. If not defined, just takes the name _logstash-input-webfetcher.log_

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java [hunk 5]
```diff
@@ -124,15 +137,13 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
 		this.crawlerUserAgent = dataMap.getString(WebFetcher.PROPERTY_CRAWLER_USER_AGENT);
 		this.crawlerReferer = dataMap.getString(WebFetcher.PROPERTY_CRAWLER_REFERER);
 
-		initializeConnection(dataMap.getString(WebFetcher.PROPERTY_PROXY_USER),
-				dataMap.getString(WebFetcher.PROPERTY_PROXY_PASS),
-				dataMap.getString(WebFetcher.PROPERTY_PROXY_HOST),
-				dataMap.getLong(WebFetcher.PROPERTY_PROXY_PORT),
-				dataMap.getBoolean(WebFetcher.PROPERTY_SSL_CHECK));
+		initializeConnection(dataMap.getString(WebFetcher.PROPERTY_PROXY_USER), dataMap.getString(WebFetcher.PROPERTY_PROXY_PASS), dataMap.getString(WebFetcher.PROPERTY_PROXY_HOST), dataMap.getLong(WebFetcher.PROPERTY_PROXY_PORT), dataMap.getBoolean(WebFetcher.PROPERTY_SSL_CHECK));
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java]: keep formatting please

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java [hunk 6]
```diff
@@ -229,16 +243,8 @@ public boolean verify(String hostname, SSLSession session) {
...
-						.connectTimeout(this.timeout.intValue())
-						.maxConnections(50)
-						.quickRender(true)
-						.blockMedia(true)
-						.userAgent(UserAgent.CHROME)
-						.logger("ch.qos.logback.core.ConsoleAppender")
-						.processes(2)
-						.loggerLevel(java.util.logging.Level.INFO)
-						.hostnameVerification(false);
+				Builder settings = Settings.builder().timezone(Timezone.EUROPE_BRUSSELS).connectTimeout(this.timeout.intValue()).maxConnections(50).quickRender(true).blockMedia(true).userAgent(UserAgent.CHROME).logger("ch.qos.logback.core.ConsoleAppender").processes(2)
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java]: keep formatting please

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java [hunk 7]
```diff
@@ -473,10 +479,7 @@ public static Result getURL(URL url, String rootUrl, Proxy proxy, Long timeout,
 
 			List<String> childPages = elements.stream().map(e -> e.attr("href"))
 					.filter(href -> (!href.startsWith(HTTP) && !href.startsWith(HTTPS) && !href.startsWith("mailto") && !href.startsWith("javascript") && !href.endsWith(".css") && !href.endsWith(".js")) || href.startsWith(HTTP + simpleUrlString) || href.startsWith(HTTPS + simpleUrlString))
-					.filter(href -> !href.equals("/") && !href.startsWith("//"))
-					.filter(href -> excludedLinkRegex.stream().noneMatch(ex -> href.matches(ex)))
-					.sorted()
-					.collect(Collectors.toList());
+					.filter(href -> !href.equals("/") && !href.startsWith("//")).filter(href -> excludedLinkRegex.stream().noneMatch(ex -> logMatch(href, ex))).sorted().collect(Collectors.toList());
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java]: keep formatting please

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/WebFetcher.java [hunk 8]
```diff
@@ -115,21 +117,21 @@ public WebFetcher(String id, Configuration config, Context context) {
...
 		jobDataMap.put(PROPERTY_PROXY_HOST, config.get(CONFIG_PROXY_HOST));
 		jobDataMap.put(PROPERTY_PROXY_PORT, config.get(CONFIG_PROXY_PORT));
 		jobDataMap.put(PROPERTY_PROXY_USER, config.get(CONFIG_PROXY_USER));
 		jobDataMap.put(PROPERTY_PROXY_PASS, config.get(CONFIG_PROXY_PASS));
+		jobDataMap.put(PROPERTY_LOG_FILE_NAME, config.get(LOG_FILE_NAME));
 
 		this.threadId = id;
 		this.urls = config.get(CONFIG_URLS).stream().map(url -> (String) url).collect(Collectors.toList());
-		this.cron = config.get(CONFIG_CRON);
+	 this.cron = config.get(CONFIG_CRON);
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/WebFetcher.java]: keep formatting please

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/WebFetcher.java [hunk 9]
```diff
@@ -141,14 +143,9 @@ public void start(Consumer<Map<String, Object>> consumer) {
 
 				String uuid = UUID.randomUUID().toString();
 
-				JobDetail job = JobBuilder.newJob(FetcherJob.class)
-						.withIdentity(uuid, GROUP_NAME)
-						.setJobData(newJobDataMap)
-						.build();
+				JobDetail job = JobBuilder.newJob(FetcherJob.class).withIdentity(uuid, GROUP_NAME).setJobData(newJobDataMap).build();
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/WebFetcher.java]: keep formatting please

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/WebFetcher.java [hunk 10]
```diff
@@ -141,14 +143,9 @@ public void start(Consumer<Map<String, Object>> consumer) {
...
-				JobDetail job = JobBuilder.newJob(FetcherJob.class)
-						.withIdentity(uuid, GROUP_NAME)
-						.setJobData(newJobDataMap)
-						.build();
+				JobDetail job = JobBuilder.newJob(FetcherJob.class).withIdentity(uuid, GROUP_NAME).setJobData(newJobDataMap).build();
 
-				Trigger trigger = TriggerBuilder.newTrigger()
-						.withIdentity(uuid, GROUP_NAME)
-						.startNow()
+				Trigger trigger = TriggerBuilder.newTrigger().withIdentity(uuid, GROUP_NAME).startNow()
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/WebFetcher.java]: keep formatting please

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/WebFetcher.java [hunk 11]
```diff
@@ -187,25 +184,8 @@ public void awaitStop() throws InterruptedException {
...
-				CONFIG_MAX_DEPTH,
-				CONFIG_TIMEOUT,
-				CONFIG_MAX_PAGES,
-				CONFIG_WAIT_JAVASCRIPT,
-				CONFIG_CRON,
-				CONFIG_THREADS,
-				CONFIG_CRAWLER_REFERER,
-				CONFIG_CRAWLER_USER_AGENT,
-				CONFIG_CHROME_DRIVER);
+		return Arrays.asList(CONFIG_URLS, CONFIG_DATA_FOLDER, CONFIG_DISABLE_SSL_CHECK, CONFIG_EXCLUDE_DATA, CONFIG_EXCLUDE_LINK, CONFIG_REFRESH_INTERVAL, CONFIG_PROXY_HOST, CONFIG_PROXY_PASS, CONFIG_PROXY_PORT, CONFIG_PROXY_USER, CONFIG_MAX_DEPTH, CONFIG_TIMEOUT, CONFIG_MAX_PAGES,
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/WebFetcher.java]: keep formatting please

File: logstash-input-webfetcher/src/main/resources/logback.xml [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+				%-5level %logger{36} - %msg%n
+			</Pattern>
+		</layout>
+	</appender>
+
+	<if condition='isDefined("elk.destination")'>
+		<then>
+
+			<appender name="LOGSTASH"
+				class="net.logstash.logback.appender.LogstashTcpSocketAppender">
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/resources/logback.xml]: do not send to LogstashTcpSocketAppender. recomended action is to leave on drive, and pick it up with filebeat

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/model/LogstashJson.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+
+/**
+ * JSON object that will be sent to logstash
+ * 
+ * @author mahytom
+ *
+ */
+@JsonInclude(Include.NON_NULL)
+public class LogstashJson implements Serializable {
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/model/LogstashJson.java]: have a look at https://github.com/logstash/logstash-logback-encoder#event-specific-custom-fields
There should be no need for a custom object

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java [hunk 14]
```diff
@@ -97,19 +102,27 @@
 	private String crawlerUserAgent;
 	private String crawlerReferer;
 	private Long maxPagesCount = 0l;
+	private Set encounteredExcludedLinks;
+	private Long excludedDataPagesCount = 0l;
+	private int urlsForDeletionCount = 0;
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java]: primitive ?

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java [hunk 15]
```diff
@@ -97,19 +102,27 @@
 	private String crawlerUserAgent;
 	private String crawlerReferer;
 	private Long maxPagesCount = 0l;
+	private Set encounteredExcludedLinks;
+	private Long excludedDataPagesCount = 0l;
```
Comment (mahytom) [inline logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java]: do we even need this ? of we send the data for analytics we can get the total inside kibana. (group by threadid)

Comment (jam-jam) [reply logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java]: Kibana is not reliable enough.

File: logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java (hunk 14 above)
Comment (jam-jam) [reply logstash-input-webfetcher/src/main/java/eu/wajja/web/fetcher/FetcherJob.java]: What's the problem with that?

**Pull request closed without merging**